From 0d5a45ff6c8d26fec0fd224e0ff8ce6bbd1e8bc3 Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Tue, 4 Oct 2022 15:01:22 +0100
Subject: x86/cpuid: Infrastructure to support pseudo feature identifiers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

A future change will want a cpuid-like identifier which doesn't have a mapping
to a feature bit.

 * Pass the feature name into the parse callback.
 * Exclude a feature value of ~0u from falling into the general set/clear bit
   paths.
 * In gen-cpuid.py, insert a placeholder to collect all the pseudo feature
   names.

No practical change.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
diff --git a/xen/arch/x86/cpuid.c b/xen/arch/x86/cpuid.c
index dc3b452e730b..a766d9ad5f82 100644
--- a/xen/arch/x86/cpuid.c
+++ b/xen/arch/x86/cpuid.c
@@ -65,7 +65,9 @@ static int __init parse_xen_cpuid(const char *s)
 
             if ( (val = parse_boolean(mid->name, s, ss)) >= 0 )
             {
-                if ( !val )
+                if ( unlikely(mid->bit == ~0u) )
+                    ;
+                else if ( !val )
                     setup_clear_cpu_cap(mid->bit);
                 else if ( mid->bit == X86_FEATURE_RDRAND &&
                           (cpuid_ecx(1) & cpufeat_mask(X86_FEATURE_RDRAND)) )
diff --git a/xen/tools/gen-cpuid.py b/xen/tools/gen-cpuid.py
index 36ea1f4a5d77..c249600362f1 100755
--- a/xen/tools/gen-cpuid.py
+++ b/xen/tools/gen-cpuid.py
@@ -294,6 +294,19 @@ def crunch_numbers(state):
         AMD_STIBP: [STIBP_ALWAYS],
     }
 
+    #
+    # Pseudo feature names.  These don't map to a feature bit, but are
+    # inserted into the values dictionary so they can be parsed and handled
+    # specially
+    #
+    pseduo_names = (
+    )
+
+    for n in pseduo_names:
+        if n in state.values:
+            raise Fail("Pseduo feature name %s aliases real feature" % (n, ))
+        state.values[n] = 0xffffffff
+
     deep_features = tuple(sorted(deps.keys()))
     state.deep_deps = {}
 
